File: src/android/app/src/main/java/org/wikimedia/epc/Sampling.java
package org.wikimedia.epc;

public class Sampling {
    private static double[] makeThresholds(double[] weights, int max_val) {
        double mv = ((double) max_val);
        double[] thresholds = new double[weights.length];
        thresholds[0] = mv * weights[0];
        for (int i = 1; i < weights.length; i++) {
            thresholds[i] = thresholds[i - 1] + (mv * weights[i]);
        }
        return thresholds;
    }
    private static double[] makeThresholds(double[] weights) {
        return makeThresholds(weights, 65535);
    }
    public static int inBucket(int rand, double[] weights, int max_val) {
        double r = (double) rand;
        if (weights.length > 1) {
            double[] thresholds = makeThresholds(weights, max_val);
            for (int i = 0; i < thresholds.length; i++) {
                if (r < thresholds[i]) {
                    return i + 1;
                }
            }
            return -1;
        } else {
            return weights.length;
        }
    }
    public static int inBucket(int rand, double[] weights) {
        return inBucket(rand, weights, 65535);
    }
    public static boolean inSample(int rand, double prob, int max_val) {
        double[] weights = {prob, 1 - prob};
        return inBucket(rand, weights, max_val) == 1;
    }
    public static boolean inSample(int rand, double prob) {
        return inSample(rand, prob, 65535);
    }
}

File: src/ios/EPC Test App/EPC Test App/EventPlatformClient/Sampling.swift
import Foundation

struct Sampling {
    static func makeThresholds(weights: [Double], _ max_val: Int = 65535) -> [Double] {
        let mv = Double(max_val)
        var segments: [Double] = weights.map() { weight in weight * mv }
        for s in 1...(segments.count - 1) {
            segments[s] += segments[s - 1]
        }
        return segments
    }
    public static func inBucket(rand: Int, weights: [Double], _ max_val: Int = 65535) -> Int {
        let thresholds = makeThresholds(weights: weights, max_val);
        let buckets = weights.count;
        let r: Double = Double(rand)
        if buckets > 1 {
            for i in 1...buckets {
                if r < thresholds[i - 1] {
                    return i
                }
            }
            return -1 // should never be called in practice
        } else {
            return buckets
        }
    }
    public static func inSample(rand: Int, prob: Double, _ max_val: Int = 65535) -> Bool {
        return inBucket(rand: rand, weights: [prob, 1 - prob], max_val) == 1;
    }
}
